Tut s ochannel

jrscott opened the pull request:
## What changes does this PR introduce?
new tutorial doc (reentrant channel)

jm-c commented:
@jrscott A side note: I tried to keep some comments in "data.diagnostics" regarding most common
parameters for each namelist. I think it could be useful since parameter from namelist are not used in the pkg/diagnostics code but copied to other that are then used (meaning, searching from then in code or header file "DIAGNOSTICS.h" will nto return anything).
In tutorial_baroclinic_gyre/input/data.diagnostics, you kept these comments in. But in this new tutorial, all these comments are gone. May I suggest to put them back ?

edoddridge commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+with a strong diabatic component, presumably due to horizontal diffusion occurring across sloping isopycnals (i.e. the so-called "Veronis effect", see
+Veronis (1975) :cite:`veronis:75` as well as other numerous papers prior to the wide-spread adoption of the GM parameterization in ocean models). [As an aside, it is for lack of a better name
+that we label this left plot of :numref:`channel_bt_MOC_res_T`, lacking either eddies or GM, as the residual circulation,
+as indeed it is identical to the Eulerian circulation in density coordinates].
```
Is this comment helpful? I'm not convinced that it is. This is the overturning circulation that the fluid feels - it is the residual overturning - the fact that it's so strongly diabatic is due to the setup.

jrscott replied:
I did not how to deal with this from a semantic perspective -- but thought it was confusing with no comment (if you were not familiar with residual circulation concept). Guessing jmarsh will weigh in on this, agree I was not satisfied in current state.

edoddridge commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+      Green dashed curves show maximum and minimum SST in each latitude band. Left plot is from non-GM run, right using GM. 
+
+Finally, let's convert the residual circulatiom shown in :numref:`channel_bt_MOC_res_T` back into depth coordinates, see :numref:`channel_bt_MOC_res_Ttoz`.
+Solid lines now display conto
```
movie is only present in the html version, but this text appears in both pdf and html.

jrscott replied:
(wrong text above but I see what you are referring to...)
Made this text less specific, applies to either movie or still (in pdf).

edoddridge replied:
Great!

edoddridge commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+      Green dashed curves show maximum and minimum SST in each latitude band. Left plot is from non-GM run, right using GM. 
+
+Finally, let's convert the residual circulatiom shown in :numref:`channel_bt_MOC_res_T` back into depth coordinates, see :numref:`channel_bt_MOC_res_Ttoz`.
+Solid lines now display conto
```
These comments are good 👍

jrscott commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+Otherwise, the numerical configuration is similar to that of tutorial :ref:`Baroclinic Ocean Gyre <tutorial_baroclinic_gyre>`).
+Potential temperature :math:`\theta` is solved prognostically,
+using Adams-Bashforth II time stepping (:numref:`adams-bashforth`, :numref:`sub_tracer_eqns_ab`).
+The fixed flux form of the momentum equations are solved, as described in :numref:`flux-form_momentum_equations`,
```
typo, fixed.

jrscott commented:
thanks for the comments @edoddridge. As you might gave expected this went through several iterations and a bunch of things you caught were not updated with changes along the way, so great to catch.

Responding to overall comments:
- yeah I should probably say more about the convection GM v. non-GM. One point was that the ML depth diag was not really reflecting the extent of cvct in non-GM in runs along the way but let me revisit.
-let me weigh the suggested changes to 4.10. In hindsight the wall on the N would have been better. In new text (not yet checked in) I more explicitly say the southern wall is equivalent to being in the north, so perhaps I could show that.  Changing up the config, or making significant changes to this figure require some work and time -- but I don't disagree with your points.
- I had the zonal mean bathy plotting at one point, not sure what I did to lose it (was not on purpose...). BTW the rational for showing this is that we are taking zonal mean layer depths so might as well show mean bathy? (even though the zonal mean bathy is a bit meaningless per se), since most of the domain is flat bottom except for the the NS ridge)

PS fwiw both jmarsh and @jm-c have reviewed the intro, but not the solution section yet. @jm-c changes to be checked in next few days.

jrscott commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+      :width: 100%
+      :align: center
+      :alt: zonal mean temp and ML depth
+      :name: channel_zm_temp_ml
```
colormap: I've reversed this colormap in the oncoming checkin (i.e. blue cold, red warm) but sounds like you feel using colormap jet in general here is to be avoided? Suggest a different colormap family for temperature maps?

mjlosch replied:
Hi @jrscott I usually use the cmocean colormaps: 
- for python: https://matplotlib.org/cmocean/

- for Matlab: https://www.mathworks.com/matlabcentral/fileexchange/57773-cmocean-perceptually-uniform-colormaps

jrscott replied:
thx, will try this

edoddridge replied:
cmocean is a great suggestion - they are excellent colour maps.

edoddridge commented:
> yeah I should probably say more about the convection GM v. non-GM. One point was that the ML depth diag was not really reflecting the extent of cvct in non-GM in runs along the way but let me revisit.

The diagnostic `convadj` is probably more useful than mld for the open ocean convection question.

> -let me weigh the suggested changes to 4.10. In hindsight the wall on the N would have been better. In new text (not yet checked in) I more explicitly say the southern wall is equivalent to being in the north, so perhaps I could show that. Changing up the config, or making significant changes to this figure require some work and time -- but I don't disagree with your points.

👍 I don't think we need to make big changes. But having a wall on one of the meridional boundaries would more accurately represent the situation.

> BTW the rational for showing this is that we are taking zonal mean layer depths so might as well show mean bathy? 

Yeah. Without the zonal-mean bathymetry, there is an unaccounted for gap. If we sum up the layer thicknesses from the bottom, then we need to start from the zonal-mean bath height, or we get silly/strange answers for the surface elevation.

jrscott commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+Munk boundary layer on our grid. However, guided by an ensemble of runs exploring parameter space,
+we found the solution with :math:`A_{h} = 100 ` m\ :sup:`2` s\ :sup:`--1`, while stable, was rather noisy.
+As a compromise, a value of :math:`A_{h} = 2000` m\ :sup:`2` s\ :sup:`--1` reduced solution noise
+whilst also controlling the strength of the barotropic current. This is the value used here.
```
jmarsh wrote that bit (improvement over what I had) but prefer your wording. Anyone else want to weigh in?

jrscott commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+
+In :numref:`reentrant_channel_soln_eddy` we will present results with
+the resolution increased by an order of magnitude, eddy-permitting. Additional required changes to the code and parameters
+are discussed in :ref:`this section <reentrant_channel_soln_eddy>`.
```
it is the same section so I just removed the latter mention/link.

jrscott commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+in years 10-30. In the mean there remains a significant heat flux into the ocean in the run without GM (solid),
+whereas with GM (dashed) the net heat uptake is also positive, but smaller. The panels
+on the right show potential temperature at the surface, mid-level (270 m) and at depth. Note in particular the warming trend at depth
+in the run without GM. The SST series display a much less obvious trend. Examining these results, we see that after 30 years our run
```
rephrased & stated explicitly that SST expected to adjust quickly

jrscott commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1167 @@
+- Package :filelink:`pkg/layers`:
+  This calculates the thickness and transport of layers of specified density (or temperature, or salinity;
+  here, temperature and density are aligned because of our simple equation of state).
+  Further explanation of :filelink:`pkg/layers` parameter options and output is given in :ref:`below <tut_SO_layers>`.
```
fixed

jrscott commented:
@edoddridge,

Regarding GM vs. non-GM comparison:
- probably should also reference Gent 2011 Ocean Modelling 39 (a perspective on GM), maybe add reference to this/mention  in intro. That being said the main purpose here is not to revisit this difference, but rather show how to use GM.
- looked at conv adj, that doesn't really show much beyond what you can infer from the MLdepth line (and as it is a 3D field, hard to show anyway). The bump in non-GM occurs I think as discussed in Gent 2011, Deacon cell brings cold water N in non-GM and make columns unstable -- however the NS ridge-notch topo creates regions where the cold water can't flow easily.
![Screen Shot 2020-03-27 at 2 14 40 PM](https://user-images.githubusercontent.com/5497358/77787998-defeb200-7036-11ea-9607-f3ec41433e26.png)
-  I don't want make to make this tutorial any longer or add figures, if possible. But I do agree more should be said about the difference in ML depth in the figure, refer to classic paper, maybe point out the odd result (without getting into this in detail) -- thinking just add a sentence in the model results where the figure discussed.

Thoughts? (or suggest sentences to add/change?)

edoddridge commented:
> probably should also reference Gent 2011 Ocean Modelling 39 (a perspective on GM), maybe add reference to this/mention in intro. That being said the main purpose here is not to revisit this difference, but rather show how to use GM.

I agree. We're not trying to do a literature review, but, I think it's a good idea to help readers learn about the gfd as well as the model. 

> looked at conv adj, that doesn't really show much beyond what you can infer from the MLdepth line (and as it is a 3D field, hard to show anyway). The bump in non-GM occurs I think as discussed in Gent 2011, Deacon cell brings cold water N in non-GM and make columns unstable -- however the NS ridge-notch topo creates regions where the cold water can't flow easily.

Again, I agree. If you know what you're looking at, then the mixed layer depth and the convective adjustment have a lot of overlap. But, the change in deep convection locations was one of the big results in the [Danabasoglu et al. (1994)](http://www.sciencemag.org/cgi/doi/10.1126/science.264.5162.1123) paper: "A comparison... shows significant improvements in the global temperature distribution, the poleward and surface heat fluxes, and the locations of deep- water formation." I therefore think it's worth mentioning, and probably worth a plot. A 2D plot like the one you posted here is sufficient, I think.

> I don't want make to make this tutorial any longer or add figures, if possible.

Absolutely. It's already very long. And fundamentally, it's about learning the model, not the gfd. Still, I think it's important to present to the convective adjustment result - it's one of the reasons why GM is so widely used. One figure, a sentence or two, and a couple of references. How do you feel about that?

jrscott commented:
@edoddridge think we are on the same page. worked up this figure, pick top or bottom set? I don't think it is bad to discuss use of convadj diag, although the top shows more info. My vote would be show top but mention convadj diagnostic.
![Screen Shot 2020-03-30 at 11 22 57 AM](https://user-images.githubusercontent.com/5497358/77930591-1197ed00-7279-11ea-8661-685e65130800.png)

jrscott commented:
@edoddridge fixing the res moc-> z coor, showing zonal mean bathy now. In the caption I changed to "and converted back into (zonal mean) depth coordinates" but perhaps that confusing
(should I say anything about the gray space at bottom and wall or is this obvious?)

edoddridge commented:
I'd go for the bottom the set - I think that the interpretation is clearer. If we show mixed layer depth, then the reader has to think through the correspondence between mld and convection. Presumably they'd get there, but, we specifically have an output that shows convection. Let's use it. As a bonus, it means we are featuring an additional diagnostic.

I also prefer the bottom set because the comparison is so stark - without GM we get convection in many open ocean cells. With GM, convection is restricted to the southernmost (and coldest) cells. It's just so clear.

> @edoddridge fixing the res moc-> z coor, showing zonal mean bathy now. In the caption I changed to "and converted back into (zonal mean) depth coordinates" but perhaps that confusing
(should I say anything about the gray space at bottom and wall or is this obvious?)

Hmm. I'm not sure. If I were writing it, I'd go for "and convert back into depth coordinates using the zonal-mean layer thicknesses summed upwards from the zonal-mean bathymetry." But, I think the parentheses work. I would mention that the grey shows the depth of the zonal-mean bathymetry. We're providing the code that produces the figures, so hopefully people can follow the logic there.

jrscott commented:
not yet addressed: any changes to schematic figure; sentence about residual circulation (in non-GM, non-eddy case) flagged by @edoddridge.

edoddridge commented:
Any progress on this @jrscott ? Would be good to merge it in.

jrscott commented:
I agree, working on getting second review completed.

jm-c commented on doc/examples/reentrant_channel/reentrant_channel.rst:
```diff
@@ -0,0 +1,1223 @@
+horizontal diffusive heat fluxes (``DFxE_TH`` and ``DFyE_TH``), and vertical diffusive heat flux (``DFrI_TH`` and ``DFrE_TH``). Note the latter is
+broken into separate implicit and explicit  components, respectively, the latter of which will only be non-zero if :filelink:`pkg/gmredi` activated.
+Although we will not examine these 3-D diagnostics below when :ref:`describing the model solution <reentrant_channel_solution>`,
+they are needed to compute a zonally-averaged meridional heat transport in any diagnostic attempt at reconciling a heat budget of the model solution.
```
Zonal advective and diffusive fluxes are not really necessary for doing a zonally-averaged heat budget or heat transport.
This might be confuse some users.

jrscott replied:
I think I accidentally dropped "or" i.e. should be "or in any diagnostic" ... but agree it is still a bit confusing. Editing in new PR.

jm-c commented:
@jrscott regarding the first point (vertical viscosity stability criteria):
I suggest to modify this paragraph to mention right away that there is no criteria that
applies since we are using implicitViscosity=T.

Instead of "Using Adams-Bashforth for vertical ...", I would suggest something like:
"Regarding the vertical viscosity, we choose to solve this term implicitly (Euler backward  
 time-stepping) by setting implicitViscosity to .TRUE. in input/data, which results in no 
 additional stability constrain on model time-step (see Section 2.6).
 Otherwise, given that our vertical resolution is quite fine near the surface (~5.m),
 the following stability criteria would have applied:
      S_lv = ... < 1.0
 limiting our choice for Av to small values."
And then continue with "For simplicity ... "

Jean-Michel

jrscott commented:
@jm-c:
1. your suggestion is good, basically followed what you said. Made last sentence in para here more clear too.
2. I've made sequences follow data.pkg order: gmredi,rbcs,layers,diagnostics throughout. Had to make a few tweaks as a sentence here and there was based on old ordering (e.g. moved sentence from data.diags section to data.layers)
3. done
4. done see other post
5. done about tau_x and SST. init temp is actually (y,z) ( I referred to the figure in the beginning) but also made this point.
6. isn't the best answer here to add diagnostics to RBCS to dump this? but I can do two things: be less negative in the phrasing :) and more specific in the last part of the sentence about computing it.

jrscott commented:
@jm-c on point 6:
This is not completely trivial what to say/do. The idea was to look at the heat gain in the sponge layer, somewhat as a sanity check what the sponge is doing. At the surface one computes relaxing heat flux as W/m2 but as this is on the sidewall, not sure whether better to compute as W or W/m2 (the former however will be affected by the varying delZ so maybe the latter is better). 
On a technical level which is the best rho to use when converting into watts here? (the Bouss rhoConst?)

jrscott commented:
actually as there are multiple gridcells thick in the sponge layer W might be better

jrscott commented:
changes in PR#373
